org.eclipse.m2e.jdt/src/org/eclipse/m2e/jdt/internal/AbstractJavaProjectConfigurator.java

```diff
@@ -845,34 +847,48 @@ private String getMinimumJavaBuildEnvironmentId(ProjectConfigurationRequest requ
     try {
       List<MojoExecution> mojoExecutions = getEnforcerMojoExecutions(request, monitor);
       for(MojoExecution mojoExecution : mojoExecutions) {
-        String version = getMinimumJavaBuildEnvironmentId(mojoExecution);
+        String version = getMinimumJavaBuildEnvironmentId(request.mavenProject(), mojoExecution, monitor);
         if(version != null) {
           return version;
         }
       }
     } catch(CoreException | InvalidVersionSpecificationException ex) {
-      log.error("Failed to determine minimum build version, assuming default", ex);
+      log.error("Failed to determine minimum build Java version, assuming default", ex);
     }
     return null;
   }
 
-  private String getMinimumJavaBuildEnvironmentId(MojoExecution mojoExecution)
-      throws InvalidVersionSpecificationException {
-    // https://maven.apache.org/enforcer/enforcer-rules/requireJavaVersion.html
-    Xpp3Dom dom = mojoExecution.getConfiguration();
-    Xpp3Dom rules = dom.getChild("rules");
-    if(rules == null) {
-      return null;
-    }
-    Xpp3Dom requireJavaVersion = rules.getChild("requireJavaVersion");
-    if(requireJavaVersion == null) {
-      return null;
-    }
-    Xpp3Dom version = requireJavaVersion.getChild("version");
-    if(version == null) {
-      return null;
+  private String getMinimumJavaBuildEnvironmentId(MavenProject mavenProject, MojoExecution mojoExecution,
+      IProgressMonitor monitor) throws InvalidVersionSpecificationException, CoreException {
+    MojoUtils mojoUtils = new MojoUtils(maven, ((MavenImpl) maven).getPlexusContainer().getContainerRealm(),
```

Comment:
@laeubi Any idea how to access the classLoader to use from outside without doing potentially invalid casting??

Reply:
It depends on what class you want to access, do you have a specific error/problem here?

Reply:
Everything works, but the pattern `((MavenImpl) maven).getPlexusContainer().getContainerRealm()` assumes that the service `IMaven` is always implemented by `MavenImpl`. Is it ok to to push `getPlexusContainer()` to IMaven?

Reply:
Actually using the plexus container should not be required at all, this is a bit an anti-pattern. Next, using the `IMaven` should be considered a potential code-smell, you are configuring a project here, so most likely you don't want the global context but a project context.

Reply:
Please have a closer look at the code. I need a ClassLoader for conversion/expression interpolation. That was the `getPlexusContainer().getContainerRealm()` before. I don't want to change everything in this PR (you can do that subsequently) but to extract the Mojo value calculation to a dedicated class this is necessary.

Reply:
> Please have a closer look at the code. I need a ClassLoader for conversion/expression interpolation.

There is a difference between "a classloader" and "the correct classloader", it seems there are two "getParameterValue" methods one uses the plugin realm and the other is using (currently) the container realm, why not use the plugin-realm in both cases? And if there is a compelling reason why use the container realm on not the maven project realm as you have a MavenProject anyways?

> but to extract the Mojo value calculation to a dedicated class this is necessary.

Just extracting code from one place to another often does not gives better results, so we always should check if the code does the right things "later" often never happens. Looking at the code I'm not sure if it is even a good diea to have this separated, a Mojo and its executions are closely coupled to a project, so maybe it would better be placed the the IProjectFacade eg like this one:
https://github.com/eclipse-m2e/m2e-core/blob/9416e758afe1a405fe6616ce110e7d974e73858e/org.eclipse.m2e.core/src/org/eclipse/m2e/core/project/IMavenProjectFacade.java#L149-L153

Reply:
Sorry but this is another issue and I don’t want to solve it in this PR. All I need is a method to retrieve nested mojo parameter values.

Reply:
Then it might be better to implement it in your plugin directly, m2e already contains some ugly hacks and we should not add another one. Using the global context and the container realm most likely giving wrong results in some setups that are hard to track.

Reply:
> in your plugin directly…

The issue and also this PR fixes an issue in m2e. But this kind of discussion doesn’t really foster contributions. Please don’t require bugfixes to improve everything. It is easy to follow up with a subsequent PR to make it a proper API. Since you committers couldn’t agree on an API please don’t let external contributors suffer from it.

Reply:
Well you asked what the proper way would be to access the classloader and I told you that the current approach is probably broken. So it does not help to "fix" broken code with some other broken code. If you think it is better fixed with separate PRs that fine. Its often the case that considerable amount of time and effort has to be done before the actual issue can be fixed, and the main goal is that it works under all circumstances we currently have.

This is not "improve everything", but these changes are already beyond trivial single one-liners and thus likely deserve closer review what might reveal related issues to be fixed for a solid soloution.

Reply:
Ok, done now.